File: Algorithm/count_complete_tree_nodes.py
class TreeNode:
    def __init__ (self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def sortedArrayToBST(self,num):
        return self.creatTree(num,0,len(num)-1)
    def creatTree(self,num,begin,end):
        if begin >end:
            return None
        mid=(begin+end)/2
        root=TreeNode(num[mid])
        root.left=self.creatTree(num,begin,mid-1)
        root.right=self.creatTree(num,mid+1,end)
        return root
    def printTree(self,root):
        if root:
            self.printTree(root.left)
            print root.val
            self.printTree(root.right)
    def countNodes(self, root):
        if not root:
            return 0
        R = root
        L = root
        h = 0
        while R:
            h += 1
            R = R.right
            L = L.left
        if not L:
            return (1<<h) -1
        return  (1<<h-1) + self.countNodes(root.left)
    def GetNodeNumKthLevel(self, root, k):
        if k == 0 or root is None:
            return 0
        if k ==1:
            return 1
        return self.GetNodeNumKthLevel(root.left, k-1) + self.GetNodeNumKthLevel(root.right, k-1)
    def countNodes1(self, root):
        def height(root):
            if not root:
                return -1
            return 1 + height(root.left)
        h = height(root)
        if h < 0:
            return 0
        if height(root.right) == h-1:
            return 2**h + self.countNodes1(root.right)
        else:
            return 2**h + self.countNodes1(root.left)

if __name__ == '__main__':
    sol = Solution()
    import random
    num=[x for x in range(15)]
    num.sort()
    print num
    
    root = sol.sortedArrayToBST(num)
    print len(num), sol.countNodes(root)

File: Algorithm/CPP/count_complete_tree_nodes.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <algorithm>
# include<time.h>
using namespace std;




struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int countNodes(TreeNode* root) {
		if (!root) return 0;
		auto p = root->left;
		auto q = root->right;
		int num = 1;
		while (q)
		{
			p = p->left;
			q = q->right;
			num = num<<1;
		}
		if (!p) return 2 * num - 1;
		return  1 + countNodes(root->left) + countNodes(root->right);
	}
	void print(TreeNode* root)
	{
		if (root)
		{
			cout << root->val << " ";
			print(root->left);
			print(root->right);
		}
	}

};

int  main()
{
	Solution sol;
	TreeNode *root1;
	root1 = new TreeNode(5);
	root1->left = new  TreeNode(4);
	root1->right = new TreeNode(8);
	//root1->left->left = new TreeNode(11);
	//root1->left->right = new TreeNode(13);
	//root1->right->left = new TreeNode(4);
	//root1->left->left->left = new TreeNode(7);
	//root1->left->left->right = new TreeNode(2);
	//root1->right->right->left =new TreeNode(5);
	//root1->right->right->right = new TreeNode(1);
	vector<int> postorder = {2, 3,1};
	vector<int> inorder = { 2, 1, 3 };
	vector<int> nums = { 1, 3, 5, 6, 7 };
	cout << sol.countNodes(root1) << endl;

	system("pause");
	return 0;
}
